dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java

```diff
@@ -60,4 +64,17 @@ public ImplementableRuleService( ProgramRuleService programRuleService )
 
     }
 
+    public Boolean hasProgramRules( Program program, String programStageUid )
+    {
+        Optional<Boolean> optionalCacheValue = getProgramRulesCache().get( program.getUid() );
+        if ( optionalCacheValue.isPresent() )
+        {
+            return optionalCacheValue.get();
+        }
+
+        boolean hasProgramRules = !getProgramRulesByActionTypes( program, programStageUid ).isEmpty();
+        getProgramRulesCache().put( program.getUid(), true );
+        return hasProgramRules;
```

Comment:
Could we potentially move this cache into the ProgramRuleEngine, where we actually fetch the program rules? Just thinking that would mean we fetch the program rules once and not twice.